src/block_ffm.rs

```diff
@@ -398,177 +317,228 @@ impl<L: OptimizerTrait + 'static> BlockTrait for BlockFFM<L> {
 
         unsafe {
             let ffm_weights = &self.weights;
-            if true {
-                _mm_prefetch(
-                    mem::transmute::<&f32, &i8>(
-                        &ffm_weights
-                            .get_unchecked(fb.ffm_buffer.get_unchecked(0).hash as usize)
-                            .weight,
-                    ),
-                    _MM_HINT_T0,
-                );
-                let field_embedding_len = self.field_embedding_len as usize;
-                let mut contra_fields: [f32; FFM_STACK_BUF_LEN] =
-                    MaybeUninit::uninit().assume_init();
+            _mm_prefetch(
+                mem::transmute::<&f32, &i8>(
+                    &ffm_weights
+                        .get_unchecked(fb.ffm_buffer.get_unchecked(0).hash as usize),
+                ),
+                _MM_HINT_T0,
+            );
+
+            /* We first prepare "contra_fields" or collapsed field embeddings, where we sum all individual feature embeddings
+               We need to be careful to:
+               - handle fields with zero features present
+               - handle values on diagonal - we want to be able to exclude self-interactions later (we pre-substract from wsum)
+               - optimize for just copying the embedding over when looking at first feature of the field, and add embeddings for the rest
+               - optimize for very common case of value of the feature being 1.0 - avoid multiplications
```

Comment:
Thinking about this, case of value being not present might also make sense to separate out of multiplying redundantly (many such ones)